hudi-common/src/main/java/org/apache/hudi/common/util/TableInternalSchemaUtils.java

```
@@ -0,0 +1,140 @@
+      List<HoodieInstant> instants = timeline.getInstants().filter(f -> f.getTimestamp().equals(String.valueOf(versionID))).collect(Collectors.toList());
+      if (instants.isEmpty()) {
+        return Option.empty();
+      }
+      byte[] data = timeline.getInstantDetails(instants.get(0)).get();
+      HoodieCommitMetadata metadata = HoodieCommitMetadata.fromBytes(data, HoodieCommitMetadata.class);
+      String latestInternalSchemaStr = metadata.getMetadata(SerDeHelper.LATESTSCHEMA);
+      return SerDeHelper.fromJson(latestInternalSchemaStr);
+    } catch (Exception e) {
+      throw new HoodieException("Failed to read schema from commit metadata", e);
+    }
+  }
+
+  public static Pair<Option<String>, Option<String>> getInternalSchemaAndAvroSchemaForClusteringAndCompaction(HoodieTableMetaClient metaClient, String compactionAndClusteringInstant) {
+    // try to load internalSchema to support Schema Evolution
+    HoodieTimeline timelineBeforeCurrentCompaction = metaClient.getActiveTimeline().filterCompletedInstants().findInstantsBefore(compactionAndClusteringInstant);
+    Option<HoodieInstant> lastInstantBeforeCurrentCompaction =  timelineBeforeCurrentCompaction.lastInstant();
+    if (lastInstantBeforeCurrentCompaction.isPresent()) {
+      try {
+        // try to find internalSchema
+        byte[] data = timelineBeforeCurrentCompaction.getInstantDetails(lastInstantBeforeCurrentCompaction.get()).get();
+        HoodieCommitMetadata metadata = HoodieCommitMetadata.fromBytes(data, HoodieCommitMetadata.class);
+        String internalSchemaStr = metadata.getMetadata(SerDeHelper.LATESTSCHEMA);
+        if (internalSchemaStr != null) {
+          String existingSchemaStr = metadata.getMetadata(HoodieCommitMetadata.SCHEMA_KEY);
+          return Pair.of(Option.of(internalSchemaStr), Option.of(existingSchemaStr));
+        }
+      } catch (Exception e) {
+        // swallow this exception
```

Why are we eating the exception ?

---

yes, we should not swallow this exception, fixed

---

this is the reason why we swallow this exception ： If we has not done some DDL, load internalSchema will failed, schema evolution should not affect the main process of current compaction.

---

hmmm, Shouldn't we be using empty schema in that case ? Can we make sure exception is thrown only during an error scenario in the first place.

---

fixed